### PR TITLE
adds default alt values for #1520

### LINKED ISSUE #1520: Images alt attribute in an editing context
Splitting this out from #1309 

One interesting thing we've recently discussed in the accessibility team, during the review of the new media widgets, is the different purpose the alt text has in the front-end and in an editing context (ie. when editing a post).

As a general rule, we all know that (I'll simplify a bit):
- purely decorative images should have an empty alt text `alt=""`
- images that convey relevant information, should use a meaningful alt text

However, when in an editing context, things are slightly different. Will try to do my best to explain. While editing a post, users need to know what the image they've added is. 
When users add an alt text to the image because they want to communicate the image purpose in the front-end, that also serves the purpose to make the image recognizable in the editor.
However, when the image is purely decorative, users will set an empty alt text. That's perfectly correct for the front-end but it will also make the image not recognizable for screen reader users because images with an `alt=""` are just ignored.

Ideally, when in the editor, also images intended to use an empty `alt=""` should have _something_ to make them perceivable and announced by screen readers. Not sure about the best option, but maybe this is something worth considering as a nice improvement Gutenberg could bring in.

As a reference, the solution implemented for the image widget was to remove the empty alt attribute in the editor when there's no alt value, so at least the filename gets announced. Not sure that's the best available option and maybe we could consider an improved solution.
https://github.com/xwp/wp-core-media-widgets/issues/47
https://github.com/xwp/wp-core-media-widgets/pull/50

### PR DESCRIPTION
**Pull request by antpb**
## Description
Sets default alt values to fix #1520 I am open to discussion around what the default value should be. I liked @afercia 's suggestion in the issue

## How to test:
1) create image block
2) add image with no alt value
3) inspect image to see that alt value is now set as a readable string "This image has an empty alt attribute, the file name is FILENAME"

### CODE REVIEW

File: packages/block-library/src/image/edit.js [hunk 1]
```diff
@@ -58,15 +58,21 @@ const LINK_DESTINATION_CUSTOM = 'custom';
...
-	let { caption } = image;
+	let { caption, alt } = image;
+	const { filename } = image;
 
 	if ( typeof caption !== 'object' ) {
 		caption = create( { html: caption } );
 	}
 
+	if ( ! alt ) {
+		alt = 'This image has an empty alt attribute, the file name is ' + filename;
```
Comment (Soean) [inline packages/block-library/src/image/edit.js]: This message should be translatable

Comment (antpb) [reply packages/block-library/src/image/edit.js]: NICE catch. Thank you.

**Review by tofumatt — CHANGES_REQUESTED** [review]
This approach works for me but as mentioned: the field should be translatable and it'd be great to add an E2E test for this (I guess in `a11y` tests or in a new `image` block test file).

File: packages/block-library/src/image/edit.js [hunk 2]
```diff
@@ -58,15 +58,21 @@ const LINK_DESTINATION_CUSTOM = 'custom';
...
-	let { caption } = image;
+	let { caption, alt } = image;
+	const { filename } = image;
 
 	if ( typeof caption !== 'object' ) {
 		caption = create( { html: caption } );
 	}
 
+	if ( ! alt ) {
+		alt = __( 'This image has an empty alt attribute, the file name is ' ) + filename;
```
Comment (tofumatt) [inline packages/block-library/src/image/edit.js]: This needs to use `sprintf()` (example usage: https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/heading/heading-toolbar.js#L18)

**Review by tofumatt — CHANGES_REQUESTED** [review]
This needs to use `sprintf()` for the translation to work.

Are you comfortable writing E2E tests for this as well? 😄

File: packages/block-library/src/image/edit.js (hunk 2 above)
Comment (tofumatt) [reply packages/block-library/src/image/edit.js]: Also, a bit pedantic but I think the comma should be a semi-colon, so it should be:

```js
alt = sprintf( __( 'This image has an empty alt attribute; its file name is "%s"' ), filename );
```

Comment (antpb) [reply packages/block-library/src/image/edit.js]: Cool. Changing. Does `caption` also need this?

Comment (tofumatt) [reply packages/block-library/src/image/edit.js]: Nope, `caption` is not content that we provide text for; it doesn't needs to be localised.

**Comment by antpb** [comment]
I'm comfortable learning how to write the tests. I bit of time reviewing the docs/source code of other tests. I think I can get this going. :)

**Comment by antpb** [comment]
Worth noting here, in my learning to build tests, I found `adding-inline-tokens.test.js` did essentially what we would want to do to test this PR. I'll also need to go through other tests to make sure there are no other image adding tests that will not have the proper alt default. 

This line is where the alt value would not have passed: https://github.com/WordPress/gutenberg/blob/master/test/e2e/specs/adding-inline-tokens.test.js#L47

@tofumatt do you think that the above test having the new default value covers this alt value test? I'm working with a new test now locally, however, other tests seem to fail when I run `npm run test-e2e` still digging into that. Figured I'd update here to make sure I'm not building duplicate tests. :)

```[1] You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
[1] For more info visit https://webpack.js.org/guides/code-splitting/
[1] npm run build exited with code 0
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gutenberg@3.9.0 pretest-e2e: `concurrently "./bin/reset-e2e-tests.sh" "npm run build"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gutenberg@3.9.0 pretest-e2e script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.```

**Comment by tofumatt** [comment]
Nice! I'll have a look sometimes today, thanks for that! 👍

**Comment by antpb** [comment]
Another thing to note that I am trying to figure out currently: Travis is not happy in comparing the raw post data because of the quotes encoding. I need to look into where that is being returned. 


Expected: `file name is "`
Got: `file name is &quot`

```  ● adding inline tokens › should insert inline image
    expect(received).toMatch(expected)
    Expected value to match:
      /<!-- wp:paragraph -->\s*<p>a\u00A0<img class="wp-image-\d+" style="width:\s*10px;?" src="[^"]+\/6c0dcb7f-627c-418b-97f8-7b563cf99b24\.png" alt="This image has an empty alt attribute; its file name is "10x10_e2e_test_image_z9T8jK.png""\/?><\/p>\s*<!-- \/wp:paragraph -->/
    Received:
      "<!-- wp:paragraph -->
    <p>a <img class=\"wp-image-153\" style=\"width: 10px;\" src=\"http://localhost:8889/wp-content/uploads/2018/10/6c0dcb7f-627c-418b-97f8-7b563cf99b24.png\" alt=\"This image has an empty alt attribute; its file name is &quot;6c0dcb7f-627c-418b-97f8-7b563cf99b24.png&quot;\"></p>
    <!-- /wp:paragraph -->"
      46 | 		// Check the content.
      47 | 		const regex = new RegExp( `<!-- wp:paragraph -->\\s*<p>a\\u00A0<img class="wp-image-\\d+" style="width:\\s*10px;?" src="[^"]+\\/${ filename }\\.png" alt="This image has an empty alt attribute; its file name is "10x10_e2e_test_image_z9T8jK.png""\\/?><\\/p>\\s*<!-- \\/wp:paragraph -->` );
    > 48 | 		expect( await getEditedPostContent() ).toMatch( regex );
         | 		                                       ^
      49 | 	} );
      50 | } );
      51 | 
```

**Comment by afercia** [comment]
Worth noting also the featured image would need the same information. The featured image preview uses a button with an aria-label, see also #10869

File: packages/editor/src/components/rich-text/core-tokens/image/index.js [hunk 3]
```diff
@@ -35,7 +35,10 @@ export const settings = {
 		);
 	},
 
-	save( { id, url, alt, width } ) {
+	save( { id, url, alt, width, filename } ) {
+		if ( ! alt ) {
+			alt = sprintf( __( 'This image has an empty alt attribute; its file name is \"%s\"' ), filename );
```
Comment (ocean90) [inline packages/editor/src/components/rich-text/core-tokens/image/index.js]: No need for the escaping.

File: packages/block-library/src/image/edit.js [hunk 4]
```diff
@@ -58,15 +58,21 @@ const LINK_DESTINATION_CUSTOM = 'custom';
...
-	let { caption } = image;
+	let { caption, alt } = image;
+	const { filename } = image;
 
 	if ( typeof caption !== 'object' ) {
 		caption = create( { html: caption } );
 	}
 
+	if ( ! alt ) {
+		alt = sprintf( __( 'This image has an empty alt attribute; its file name is \"%s\"' ), filename );
```
Comment (ocean90) [inline packages/block-library/src/image/edit.js]: No need for the escaping.

File: test/e2e/specs/adding-inline-tokens.test.js [hunk 5]
```diff
@@ -44,7 +44,7 @@ describe( 'adding inline tokens', () => {
 		await page.click( '.media-modal button.media-button-select' );
 
 		// Check the content.
-		const regex = new RegExp( `<!-- wp:paragraph -->\\s*<p>a\\u00A0<img class="wp-image-\\d+" style="width:\\s*10px;?" src="[^"]+\\/${ filename }\\.png" alt=""\\/?><\\/p>\\s*<!-- \\/wp:paragraph -->` );
+		const regex = new RegExp( `<!-- wp:paragraph -->\\s*<p>a\\u00A0<img class="wp-image-\\d+" style="width:\\s*10px;?" src="[^"]+\\/${ filename }\\.png" alt="This image has an empty alt attribute; its file name is "10x10_e2e_test_image_z9T8jK.png""\\/?><\\/p>\\s*<!-- \\/wp:paragraph -->` );
```
Comment (ocean90) [inline test/e2e/specs/adding-inline-tokens.test.js]: You have to use the HTML entity `&quot;` for the quotes to check for valid HTML.

**Review by ocean90 — CHANGES_REQUESTED** [review]

**Comment by antpb** [comment]
Closing this as I have rebased and made all of the above code change recommendations at his PR: https://github.com/WordPress/gutenberg/pull/10960. As for @afercia 's recent comment about featured image, I need to look into that. Might make a new issue so we can get these default alt values added asap. I have a PR already open that is manipulating featured image behavior. I'll see if I can just wrap the default alt values into there.

**Pull request closed without merging**